### ISSUE TITLE
[Tasks]: Make two modules of wofost72_pp differentiable

### ISSUE THREAD

**Issue by SarahAlidoost** [WUR-AI/diffWOFOST#8]
As a part of building the prototype, we want to make [`Root_Dynamics`](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/root_dynamics.py) and [`Leaf_Dynamics`](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/leaf_dynamics.py) differentiable w.r.t `"SPAN" and "TDWI"`. To do, check for look-up tables, hard thresholds, and mathematical operations on arrays​. The module [`Wofost72`](https://github.com/ajwdewit/pcse/blob/master/pcse/crop/wofost72.py) also needs refactoring to use `torch.nn.Module`. 

The current structure of PCSE regarding `Wofost72_PP`:
- [class Wofost72_PP(Engine)](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/models.py#L8)
    - └── [class Engine(BaseEngine)](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/engine.py#L33)
        - └── [class Wofost72(SimulationObject)](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/crop/wofost72.py#L26C1-L26C33)
            - └── [9 modules](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/crop/wofost72.py#L35-L43)        
        - └── [repository WOFOST_crop_parameters: branch wofost72](https://github.com/ajwdewit/WOFOST_crop_parameters/tree/wofost72)
        
Sub-tasks:
- [ ] Implement a differentiable version of `Root_Dynamics`:
    - [ ] Identify operations on "SPAN" and "TDWI" and ensure all math uses PyTorch tensors
    - [ ] Wrap the model in `torch.nn.Module` with `forward()` method
- [ ] Implement a differentiable version of `Leaf_Dynamics` :
    - [ ] Identify operations on "SPAN" and "TDWI" and ensure all math uses PyTorch tensors
    - [ ] Wrap the model in `torch.nn.Module` with `forward()` method
- [ ] Identify the gradient path of `Root_Dynamics` and `Leaf_Dynamics`  in  `Wofost72`:
    - [ ] Wrap the model in `torch.nn.Module` with `forward()` method
    - [ ] Identify operations on "SPAN" and "TDWI" and ensure all math uses PyTorch tensors
- [ ] Check `Engine` and `wofost72_pp` classes: They don't need to be wrapped by `torch.nn.Module`, but identify how they influence the gradients path w.r.t those two parameters. 
- [ ] Build the framework (see the example structure in #3):
    - [ ] Implement a dummy `MLModel`: use `torch.nn.Module` with `forward()` method
    - [ ] Implement the `PhysicalModel`: plug in differentiable wofost72_pp
    - [ ] Implement the `HybridModel`: use `torch.nn.Module` with `forward()` method
- [ ] Implement tests and checks: 
    - [ ] Unit test each module, see some [test data](https://github.com/ajwdewit/pcse/tree/master/tests/test_data)
    - [ ] Confirm gradients exist
- [ ] Create training module including loss functions and optimizers

**Note:** To avoid copying-pasting the other classes and modules, you may use the patching technique e.g. `from unittest.mock import patch` when needed.

**Comment by SarahAlidoost** [comment]
>1. If I understand correctly, each "module" in wofost uses some ParameterTemplate that gets initialized using values provided by a single ParameterProvider shared by all "modules". Do you know why parameters are not simply initialized within the ParameterProvider? Having multiple instances of the same parameter value will now also result in multiple gradients being computed per parameter, but each instance will be updated separately. Not sure if this is desirable/intended.

In PCSE, the `ParameterProvider` is basically a mutable, dictionary-like object. Think of it as a container that stores parameter names and their values. To create a `ParameterProvider`, we can:
```python
from pcse.base.parameter_providers import ParameterProvider
param_provider = ParameterProvider()
param_provider
<pcse.base.parameter_providers.ParameterProvider object at 0x7b7bbef50050>
```
Of course, we’ll usually want to give it something useful, like crop parameters:
```python
param_provider = ParameterProvider(cropdata=crop_parameters)
```
At this point, what we have is just a container for parameters (their names and values). These values can be anything: floats, tensors, etc. Why PCSE use ParameterProvider? The idea is explained [here](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/pcse/base/parameter_providers.py#L24-L33).

Now, each crop module is a Python class, and it defines a class attribute called `Parameters`. This attribute specifies makes sure the crop has everything it requires before running a simulation. Every crop module is also a subclass of `SimulationObject`. When you create a crop module instance, the `__init__` method of `SimulationObject` kicks in. That method calls the crop module’s `initialize() ` function. Inside `initialize()`, there is something like this: `self.params = self.Parameters(parvalues)`. Here’s what’s happening:

- `self.Parameters` knows which parameters are required by the crop module.
- `parvalues` is the dictionary (from the `ParameterProvider`) that actually holds all the parameter values.
- The required parameters get pulled out and stored in `self.params`.

Now, regarding your comment: “_Having multiple instances of the same parameter value will now also result in multiple gradients being computed per parameter._” Let’s take the parameter `TDWI` as an example, which is used in two crop modules: `leaf_dynamics` and `root_dynamics`. Suppose we define it like this: `tdwi = torch.nn.Parameter(torch.tensor(0.2, dtype=torch.float64))` and then pass it to both modules. Inside the modules, `root_dynamics.params["TDWI"]` and `leaf_dynamics.params["TDWI"]` are actually pointing to the same underlying tensor in memory. When PyTorch builds the computation graph, it only tracks the tensor objects themselves. So in this case, both `root_dynamics.params["TDWI"]` and `leaf_dynamics.params["TDWI"]` are just two aliases referring to the same tensor. That means only one set of gradients will be computed for that tensor, not separate ones.

That said, we know there are a few technical issues with `ParameterProvider` (like #20 and #21). I think they mostly affect efficiency rather than differentiability. Still, we don’t fully know all the tricky interactions that might pop up in WOFOST when multiple modules work together. I might also miss something in your question. So, if you have an example where things go wrong with the differentiability of outputs w.r.t parameters, please share it with us. Thanks!

**Comment by SarahAlidoost** [comment]
> 2.How I understand the project currently is that we end up with a "DifferentiableWofost" class that inherits from torch.nn.Module. A "DifferentiableWofost" object contains some ParameterProvider object which I think should probably also be a torch.nn.Module. In that case the parameter instances are also shared between wofost modules and it makes it easier to implement ParameterProvider objects based on machine learning models -- something I suspect will be a common use-case.

 Issue #3 shows the idea behind hybrid modeling: an ML model predicts some physical parameters, which are then fed into a physics-based model; it can be WOFOST or any other crop module, so the two models (ML and physics) are combined in a hybrid architecture, see the chain rule in #3. Yes, this means the physical model needs to be “PyTorch-friendly”: it should be wrapped in `nn.Module` and have proper `__init__` and `forward` functions, just like in [this test example](https://github.com/WUR-AI/diffWOFOST/blob/468d89cdfef52cbabdd7ba6bb102814e2f765da9/tests/physical_models/crop/test_leaf_dynamics.py#L65-L99).

Any parameter for which we want gradients computed should be wrapped with `torch.nn.Parameter`. Not the `ParameterProvider` itself, that’s just a container, but the actual parameters inside it should be `torch.nn.Parameter`, like in [this example](https://github.com/WUR-AI/diffWOFOST/blob/468d89cdfef52cbabdd7ba6bb102814e2f765da9/tests/physical_models/crop/test_leaf_dynamics.py#L207). For differentiability to work, all the math operations need to be tensor-based and differentiable, which is why variables like parameters, state variables, or rates are wrapped as `torch.tensor`.  Hope this addresses your concerns. Otherwise, please let me know.

**Comment by SarahAlidoost** [comment]
> 4.In the differentiable wofost WOFOST_Leaf_Dynamics.integrate function, why are state variables detached from the gradient computation graph? Does this not affect gradient computation wrt model parameters?

Good catch, thanks! 👍 I fixed it. It seems that with `.detach().clone()`, the gradients w.r.t the parameter are no longer correct, even though forward/backward still run without errors 🤔 . After removing `.detach()`, the gradients match the numerical check. I also noticed while debugging that, with regards to the test data [`test_leafdynamics_wofost72_01.yaml`](https://github.com/WUR-AI/diffWOFOST/blob/leaf_dynamics/tests/physical_models/test_data/test_leafdynamics_wofost72_01.yaml), parameter `SPAN` doesn’t seem to have any effect on output `TWLV` (I plotted the two to check). So before setting `requires_grad=True` on a parameter, it’s probably a good idea to confirm that the parameter actually influences the output. Not sure yet if this is just for the specific test case `test_leafdynamics_wofost72_01.yaml` or something more general in `leaf_dynamics`. What do you think?

**Comment by SarahAlidoost** [comment]
5. Imo it is probably good to use a single dtype across the entire differentiable wofost model instead of specifying it per module. I can imagine this was done to just create an initial demo though. I think PyTorch conventionally specifies the initialization of nn.Module objects like DifferentiableWofost(..., dtype=None, device=None) where the specified dtype and device are used to initialize all tensors and uses the torch defaults otherwise. Should we follow this convention? Also when creating new tensors in a function call (e.g. WOFOST_Leaf_Dynamics.calc_rates) its probably good to infer the dtype and device from existing tensors (like parameters that were already specified).

Good suggestions 👍  I opened issue #23, so when start looking into efficiency, we’ll address it. These should definitely be exposed so they can be adjusted from outside the code if needed.

**Comment by SarahAlidoost** [comment]
> 6.I saw you used a sharpness parameter in the sigmoid/soft threshold function. Do you know if there's a go-to approach for determining this value? I can imagine it's also quite dependent on the magnitude of the input.

Good question! For now, I just changed it from a small value (10) to 100, then 1000, and checked whether the tests still passed. With `sharpness = 1000.0`, the sigmoid behaves almost like a step function. But I’m not sure if this works generally or if it’s just fine for the current test data [test_leafdynamics_wofost72_01.yaml](https://github.com/WUR-AI/diffWOFOST/blob/leaf_dynamics/tests/physical_models/test_data/test_leafdynamics_wofost72_01.yaml). I also opened an issue #24 so we don’t forget about it. I also don’t know how many similar approximations we’ll need in other modules. One option would be to run all test data and try to find an optimal sharpness value. Another option would be to make it a learnable parameter. 
The key point is that these approximation will change the results. How to validate them? In one of the update meetings, I discussed this issue and the question of the accuracy expectations for diffwofost compared to wofost. For example, should we consider tests passed if results are within, say, 10% relative difference? More? Less? what do you think?

**Comment by ronvree** [comment]
Thanks for the quick and extensive replies! I'll post my comments one by one

1. I was not 100% familiar with the exact interactions of `Parameters`, `ParameterProvider`, `ParameterTemplate` and the traitlets library and how they initialize parameters. I was mostly just worried that a 1:1 conversion would create hard copies of parameters and not shallow ones. I think you are right that these issues will not occur and either way you are very aware of the issue :)  Thanks!

2. Very clear, thanks! 

3. I should have been more clear here! Looking at [#3](https://github.com/WUR-AI/diffWOFOST/issues/3), I see that with every forward call a new `Wofost72_PP` object is created. The optimization of the model parameters typically involves a lot of simulations/forward calls. I am not completely familiar with what happens behind the screens in PCSE, but my worry is that with the creation of every new `Wofost72_PP` object a lot of expensive operations will take place that are not relevant to the simulation outcome (e.g. reading/writing config files). I think we should verify whether this is actually the case and if so, consider the options of either removing these operations, re-use a `Wofost72_PP` object (by updating its internal state), or create a new object that only mimics the crop physiology. 

4. Nice! What will happen is that after calling `.detach()` PyTorch will just treat the states as constants when computing gradients. I think the final state was probably still dependent in some way on the parameters (so did not get the `.detach` call), and then used to compute the gradients. So no error will occur with the forward/backward calls, but the gradients just might be incorrect. How do you verify the gradients currently? Also, I'm not sure, but I would imagine it is difficult to provide guarantees about which parameter doesn't affect which output and there isn't a big downside to having some "idle" parameters afaik. 
 
5. Sounds good!

6. Tbh I'm not sure! There is indeed a trade-off between model accuracy and "differentiability", in the sense that closer approximations provide "weaker" gradients. To be continued :)

Thanks again!

**Comment by SarahAlidoost** [comment]
> 3. I should have been more clear here! Looking at [#3](https://github.com/WUR-AI/diffWOFOST/issues/3), I see that with every forward call a new `Wofost72_PP` object is created. The optimization of the model parameters typically involves a lot of simulations/forward calls. I am not completely familiar with what happens behind the screens in PCSE, but my worry is that with the creation of every new `Wofost72_PP` object a lot of expensive operations will take place that are not relevant to the simulation outcome (e.g. reading/writing config files). I think we should verify whether this is actually the case and if so, consider the options of either removing these operations, re-use a `Wofost72_PP` object (by updating its internal state), or create a new object that only mimics the crop physiology.

You’re right that in the proposed hybrid modeling structure in #3, the line `wofost = Wofost72_PP(params, …)` inside the `forward` function creates a new instance of `Wofost72_PP` with each forward call. But, this refers to a new implementation of `Wofost72_PP` that will include a differentiable `wofost_72` and a more efficient engine, and not to the current PCSE version. In the proposed structure, the differentiable `wofost_72` will connect all the differentiable crop modules, and as you suggested, the new engine should be able to update everything needed during each forward call but not the crop model's state. 

On data handling: loading data happens in the engine, not in the physical model (e.g. wofost_72). But loading into the engine doesn’t necessarily mean “reading” data. Reading is handled by components like `ParameterProvider`, `WeatherDataProvider`, and possibly `AgroManager`. Only `ConfigurationLoader` currently can’t accept a config object (we’re aware of this, see #22). Also, there’s no file writing in the current PCSE engine. There is an option to save outputs, but that only stores them in memory for later access once the model run finishes.

To get more familiar with the PCSE engine, I’d suggest checking the `setUp` function in [the PCSE test file](https://github.com/ajwdewit/pcse/blob/6c22e0345f71338db085c4159dc5fe2463ea1972/tests/run_tests.py#L60). It’s a minimal example showing where data is read and is passed to the engine, and later the engine runs a crop model. In the test, the crop model can be either Wofost72_PP or any of the individual crop modules used within Wofost72_PP.

You mentioned that “_a lot of expensive operations will take place that are not relevant to the simulation outcome (e.g. reading/writing config files)_.” two points on this:

1. As far as I know, the only file-related operation is reading the config file in the engine using `ConfigurationLoader`. I’m not aware of any writing config file, can you please explain?
2. Could you clarify what other expensive operations you’ve noticed? Was the issue mainly runtime, memory usage, or something else? We can also run a profiler on the engine to find any bottlenecks. I submitted issue #30.

**Comment by SarahAlidoost** [comment]
> 4. Nice! What will happen is that after calling `.detach()` PyTorch will just treat the states as constants when computing gradients. I think the final state was probably still dependent in some way on the parameters (so did not get the `.detach` call), and then used to compute the gradients. So no error will occur with the forward/backward calls, but the gradients just might be incorrect. How do you verify the gradients currently? Also, I'm not sure, but I would imagine it is difficult to provide guarantees about which parameter doesn't affect which output and there isn't a big downside to having some "idle" parameters afaik.

To answer your question _"How do you verify the gradients currently?"_, I calculate numerical gradients using "the central difference approximation" see [this function](https://github.com/WUR-AI/diffWOFOST/blob/35a9ed09d398ed9ddd8596cff55ba7677fe5bf49/tests/physical_models/crop/test_leaf_dynamics.py#L101), and check whether it is equal to the gradients calculated by pytorch with an absolute tolerance of 0.001. For example, [see this test](https://github.com/WUR-AI/diffWOFOST/blob/35a9ed09d398ed9ddd8596cff55ba7677fe5bf49/tests/physical_models/crop/test_leaf_dynamics.py#L225).
